### DISCUSSION TITLE
A way to simulate "Bilateral blur"

### DISCUSSION THREAD

**Discussion by F-Burning** [opentoonz/opentoonz#6813] [Ideas]
### Ask a question related to OpenToonz

Adobe has a type of blur called bilateral blur, which allows you to soften the areas of an image, but respecting the edges, giving a result like watercolor, like in this example:

![04_00000](https://user-images.githubusercontent.com/74098734/208135618-69a16c36-2637-4447-844a-087f417a956c.png)

Here is the picture without the effect:

![001_00000](https://user-images.githubusercontent.com/74098734/208135963-9e49df2a-508a-470d-a9ec-472dbc9a1d0c.png)

If I'm not mistaken, this type of blur is not found in the program, and I was wondering if there was a way to emulate it.

**Comment by RodneyBaker** [comment]
It might be pretty direct to implement a bilateral (blur) shader.
One example here:  https://www.shadertoy.com/view/4dfGDH

Interactive example:  http://vcg.isti.cnr.it/Publications/2012/BCCS12/

**Comment by RodneyBaker** [comment]
From what I can tell, FFMPEG does not have a true Bilateral Blur although Image Magick reportedly does.

FFMPEG does have some approximations such as:

Block-matching 3D denoising algorithm:

`ffmpeg -i input.png -vf "bm3d=sigma=3" output.png`

Approximation via non local means (nlMeans):

`ffmpeg -i input.png -vf "nlmeans=s=2:r=7:p=3" output.png`

Interestingly, it seems we can call Image Magick FROM FFmpeg like this:

`ffmpeg -i input.png -f image2pipe -vcodec ppm - | magick ppm:- -bilateral-blur 10x10+5+3 png:- | ffmpeg -i - output.png`

That's if we want to get Image Magick's exact approach.

If our FFMPEG is built with OpenCV support we can use OpenCV:

`ffmpeg -i input.png   -vf opencv=bilateralFilter:diameter=9:sigmaColor=75:sigmaSpace=75" output.png`

Most users would not be using a copy of FFMPEG with OpenCV built in and most users would not have Image Magick so other approaches likely are necessary.

**Comment by RodneyBaker** [comment]
FFMpeg may have some other options.  For example:  `ffmpeg -i in.png -vf "format=yuv444p,bilateral=sigmaS=3:sigmaR=0.1" out.png`

I haven't tested in that direction yet as I was shifting to see if a Shader might be an option.
